File: sdks/go/extensions/dispute/facilitator.go
package dispute

import "time"

// FacilitatorResolveInput holds inputs for the arbiter-side handler.
type FacilitatorResolveInput struct {
	DisputeHash       string
	Verdict           Verdict
	SettledAmount     string
	RefundTransaction string
	Version           int   // 0 = default to 1
	IssuedAt          int64 // 0 = time.Now().Unix()
}

// FacilitatorHandler signs resolutions for the "facilitator" arbiterScheme.
// The arbiter address is the handler's signer address.
type FacilitatorHandler struct {
	signer Signer
}

// NewFacilitatorHandler builds a handler that uses the given signer as
// arbiter. The handler's GetArbiterAddress is the signer's address; the
// signed resolution's payload.arbiter is filled accordingly.
func NewFacilitatorHandler(signer Signer) *FacilitatorHandler {
	return &FacilitatorHandler{signer: signer}
}

// GetArbiterAddress returns the address this handler uses as arbiter.
func (h *FacilitatorHandler) GetArbiterAddress() string {
	return h.signer.GetAddress()
}

// ResolveDispute signs a resolution for a verified dispute.
func (h *FacilitatorHandler) ResolveDispute(input FacilitatorResolveInput) (*SignedResolution, error) {
	version := input.Version
	if version == 0 {
		version = 1
	}
	issuedAt := input.IssuedAt
	if issuedAt == 0 {
		issuedAt = time.Now().Unix()
	}
	payload := &ResolutionPayload{
		Version:           version,
		DisputeHash:       input.DisputeHash,
		Verdict:           input.Verdict,
		SettledAmount:     input.SettledAmount,
		Arbiter:           h.GetArbiterAddress(),
		IssuedAt:          issuedAt,
		RefundTransaction: input.RefundTransaction,
	}
	return CreateSignedResolution(h.signer, payload)
}

// BuildFacilitatorResolution is a one-call helper for facilitators
// acting on a pre-decided verdict.
func BuildFacilitatorResolution(
	handler *FacilitatorHandler,
	disputeHash string,
	verdict Verdict,
	settledAmount string,
	refundTransaction string,
) (*SignedResolution, error) {
	return handler.ResolveDispute(FacilitatorResolveInput{
		DisputeHash:       disputeHash,
		Verdict:           verdict,
		SettledAmount:     settledAmount,
		RefundTransaction: refundTransaction,
	})
}

File: sdks/typescript/packages/extensions/src/dispute/facilitator.ts
/**
 * Facilitator-side helpers for the Dispute extension.
 *
 * Provides the integration glue for facilitators that serve as the
 * `facilitator` arbiterScheme (the t402 facilitator acts as arbiter for
 * disputes). For `contract` or `external` arbiterSchemes, the resolution
 * is signed elsewhere and the facilitator only proxies / verifies.
 */

import type {
  DisputeVerdict,
  EIP712DisputeSigner,
  ResolutionPayload,
  SignedDispute,
  SignedResolution,
} from "./types";
import { createSignedResolution } from "./signing";

/** Inputs to the facilitator-as-arbiter resolution builder. */
export interface FacilitatorResolveInput {
  /** EIP-712 hash of the SignedDispute being resolved. */
  disputeHash: string;
  /** Arbiter verdict (one of upheld_full/upheld_partial/denied/void). */
  verdict: DisputeVerdict;
  /** Settled refund amount; MUST be 0 for denied/void. */
  settledAmount: string;
  /** On-chain or off-chain refund reference, optional. */
  refundTransaction?: string;
  /** Resolution version, defaults to 1. */
  version?: number;
  /** Issued-at unix-seconds; defaults to current time. */
  issuedAt?: number;
}

/**
 * Facilitator handler that produces a SignedResolution from a verdict
 * decision. The facilitator's signing key is the arbiter address; that
 * address MUST match what the merchant advertised in the 402 dispute
 * terms (`arbiter` field).
 */
export interface DisputeFacilitatorHandler {
  /** Resolve a verified dispute by signing a resolution payload. */
  resolveDispute(input: FacilitatorResolveInput): Promise<SignedResolution>;
  /** The arbiter address (= the facilitator's signing key). */
  getArbiterAddress(): string;
}

/**
 * Build a DisputeFacilitatorHandler from a generic EIP-712 dispute
 * signer (typically the facilitator's wallet).
 *
 * @param signer The EIP-712 signer that holds the facilitator's key.
 */
export function createDisputeFacilitatorHandler(
  signer: EIP712DisputeSigner,
): DisputeFacilitatorHandler {
  const arbiterAddress = signer.getAddress();
  return {
    getArbiterAddress() {
      return arbiterAddress;
    },
    async resolveDispute(input) {
      const payload: ResolutionPayload = {
        version: input.version ?? 1,
        disputeHash: input.disputeHash,
        verdict: input.verdict,
        settledAmount: input.settledAmount,
        arbiter: arbiterAddress,
        issuedAt: input.issuedAt ?? Math.floor(Date.now() / 1000),
        ...(input.refundTransaction !== undefined
          ? { refundTransaction: input.refundTransaction }
          : {}),
      };
      return createSignedResolution(signer, payload);
    },
  };
}

/**
 * Helper for facilitators: given a SignedDispute and a policy verdict,
 * produce the resolution in one call. Useful when the verdict has been
 * pre-decided by an off-chain arbitration process.
 *
 * @param handler The facilitator handler.
 * @param dispute The signed dispute being resolved (used for disputeHash
 *   computation; caller is responsible for providing the hash separately).
 * @param disputeHash EIP-712 hash of the signed dispute.
 * @param verdict The verdict to issue.
 * @param settledAmount The refund amount (must satisfy the verdict↔amount
 *   consistency rule — see specs/extensions/dispute.md §Verification).
 * @param refundTransaction Optional on-chain or off-chain refund reference.
 */
export async function buildFacilitatorResolution(
  handler: DisputeFacilitatorHandler,
  _dispute: SignedDispute,
  disputeHash: string,
  verdict: DisputeVerdict,
  settledAmount: string,
  refundTransaction?: string,
): Promise<SignedResolution> {
  return handler.resolveDispute({
    disputeHash,
    verdict,
    settledAmount,
    ...(refundTransaction !== undefined ? { refundTransaction } : {}),
  });
}
